addons/mail/static/src/js/mail_mention.js

```diff
@@ -0,0 +1,225 @@
+                _.each(res, function(r) {self.partners[r.id] = r;});
+                self.search_keys[self.search_string] = _.pluck(res, "id");
+                self.show_dropdown();
+            });
+            return;
+        }
+        self.show_dropdown();
+    },
+    find_n_test: function(){
+        var self = this;
+        var test_string = function(search_str){
+             var pattern = /(^@|(^\s@))/g;
+             var regex_start = new RegExp(pattern);
+             if (regex_start.test(search_str) && search_str.length >= self.min_charactor ){
+                 search_str = _.str.ltrim(search_str.replace(pattern, ''));
+                 return search_str.indexOf(' ') < 0 &&  _.isNull(/\r|\n/.exec(search_str))? search_str:'';
+             }
+            return '';
+        };
+        var desctiption = this.$textarea.val(),
+        left_string = desctiption.substring(0, this.get_cursor_position(this.$textarea)),
+        search_str = desctiption.substring(left_string.lastIndexOf("@") - 1, this.get_cursor_position(this.$textarea));
+        return test_string(search_str);
+    },
+    keyup: function(e) {
+        var self = this;
+        if(_.contains(this.keyup_list, e.which)) { return;}
+        
+        this.recent_string = this.find_n_test();
+        if(this.recent_string){
+            this.search_string = this.recent_string;
+            clearTimeout(this.timer);
+            this.timer = setTimeout(function() {
+                self.search_n_store();
+            }, this.typing_speed);
+        }else if(this.$dropdown.find("li")){
+            this.clear_dropdown();
+        }
+    },
+
+    show_form_popup: function(email){
+        var deferred = $.Deferred();
+        if(email){
+            deferred.resolve();
+            return deferred;
+        }
+        var partner_form_popup = new form_common.FormOpenPopup(this);
+            partner_form_popup.show_element(
+                'res.partner',
+                id,
+                {
+                    'force_email': true,
+                    'ref': "compound_context",
+                    'default_name': name
+                },
+                {
+                    title: _t("Please complete partner's informations"),
+                    write_function: function(id, data, options) {
+                        return this._super(id, data, options).done(function() {
+                            if(data.name)
+                                self.partners[id]['name'] = data.name;
+                            self.partners[id]['email'] = data.email;
+                            deferred.resolve();
+                        });
+                    },
+                }
+            );
+            partner_form_popup.on('closed', self, function() {
+                self.$textarea.focus();
+            });
+        return deferred;
+    },
+    find_n_replace: function(id) {
+        var self = this;
+        var partner = this.partners[id],
```

Comment:
It will crash when there is no partner in this.partners. it seems partners list not maintain properly . for example @arg display list of three partner I select any one of it them then I delete it manually (by backspace), then it will not show that partner again in selection list